### PR TITLE
add callback to pubsub.unsubscribe and test (#300)

### PR DESCRIPTION
**Pull request by noot**
As per the issue I opened: https://github.com/libp2p/js-libp2p/issues/300

Upon `pubsub.unsubscribe`, the node was successfully unsubscribing, however the callback was never being called, so there was no explicit way to confirm success.

I have changed `unsubscribe` as to call the callback, so it now complies with the interface-ipfs-core spec. I have also added a test.

Let me know if any changes are needed, thanks!

### CODE REVIEW

**Comment by noot** [comment]
Not sure how to fix the remaining failing tests, sorry.

File: src/pubsub.js [hunk 1]
```diff
@@ -43,6 +43,8 @@ module.exports = (node) => {
       if (floodSub.listenerCount(topic) === 0) {
         floodSub.unsubscribe(topic)
       }
+
+      setImmediate(() => callback())
```
Comment (jacobheun) [inline src/pubsub.js]: Since the callback is being added, even though it should have already exists, we should check that it is actually a function before invoking it.

File: test/pubsub.node.js [hunk 2]
```diff
@@ -52,23 +52,26 @@ function stopTwo (nodes, callback) {
...
-                expect(err).to.not.exist()
-              }), 500)
-            }
-          )
+            }), 500)
+            setTimeout(() => nodes[0].pubsub.unsubscribe('pubsub', handler, (err) => {
+              expect(err).to.not.exist()
+              console.log('tunsubscribed!')
+              done()
+            }), 600)
```
Comment (jacobheun) [inline test/pubsub.node.js]: We should get rid of the timeouts in this test. Subscribing -> publishing -> unsubscribing should be able to be executed in serial successfully without a timeout.

Comment (noot) [reply test/pubsub.node.js]: done!

File: src/pubsub.js (hunk 1 above)
Comment (noot) [reply src/pubsub.js]: makes sense, I added a check

File: test/pubsub.node.js [hunk 3]
```diff
@@ -52,23 +52,26 @@ function stopTwo (nodes, callback) {
...
-              setTimeout(() => nodes[1].pubsub.publish('pubsub', data, (err) => {
-                expect(err).to.not.exist()
-              }), 500)
-            }
-          )
+            })
+            nodes[0].pubsub.unsubscribe('pubsub', handler, (err) => {
+              expect(err).to.not.exist()
+              console.log('\tunsubscribed!')
+              done()
```
Comment (jacobheun) [inline test/pubsub.node.js]: It's possible done could be called before we run the expect in the publish handler. It would be good to include `chai-checkmark` in this file and mark each of the expects.  That way we can do `expect(4).checks(done)` so we can ensure everything is verified before finishing the test.

Remove the console.log as well and I think this is good to go. Thank you for the PR!

Comment (noot) [reply test/pubsub.node.js]: Sounds good! I've added chai-checkmark and the check to make sure all the expects happen, as well as moved done and got rid of the console.log.

Comment (noot) [reply test/pubsub.node.js]: When I remove `done()` at line 72, I get `Error: Timeout of 2000ms exceeded. For async tests and hooks, ensure "done()" is called; if returning a Promise, ensure it resolves.` If we place `done()` at line 77, at the end of the waterfall, it works. I'm not sure if this is corrrectly allows for the expect check to happen though.

**Comment by jacobheun** [comment]
Sorry for the delay, I've been away on holiday but am back now. The test done logic is still not happening correctly. Done is being called and is passed as a callback to waterfall, this isn't correct. The timeout you were hitting is probably because things aren't firing in the right order, or are being triggered early. I can take a deeper look at what the issue is later today.

**Comment by noot** [comment]
@jacobheun No worries! Sounds good, thanks. Let me know what issues you find.

**Comment by jacobheun** [comment]
There were some errors in the test, primarily some of the function arguments and `.mark()` not being added for chai-checkmark. The delay after the subscribe is also important, otherwise publish will likely be called before the peer is aware of the subscriber.

I think it was a bit hard to follow so I made some changes along with the fixes that hopefully make things a bit easier to read.

```js
    it('start two nodes and send one message, then unsubscribe', (done) => {
      // Check the final series error, and the publish handler
      expect(2).checks(done)

      let nodes
      const data = Buffer.from('test')
      const handler = (msg) => {
        // verify the data is correct and mark the expect
        expect(msg.data).to.eql(data).mark()
      }

      series([
        // Start the nodes
        (cb) => startTwo((err, _nodes) => {
          nodes = _nodes
          cb(err)
        }),
        // subscribe on the first
        (cb) => nodes[0].pubsub.subscribe('pubsub', handler, cb),
        // Wait a moment before publishing
        (cb) => setTimeout(cb, 500),
        // publish on the second
        (cb) => nodes[1].pubsub.publish('pubsub', data, cb),
        // unsubscribe on the first
        (cb) => nodes[0].pubsub.unsubscribe('pubsub', handler, cb),
        // Stop both nodes
        (cb) => stopTwo(nodes, cb)
      ], (err) => {
        // Verify there was no error, and mark the expect
        expect(err).to.not.exist().mark()
      })
    })
```

If that looks good and makes sense to you can you push the update?

**Comment by noot** [comment]
@jacobheun That's a lot easier to follow, I think it makes sense.  I'll push the changes now

**Comment by jacobheun** [comment]
I'm going to look into the windows ci failure. It's possible it's just flakey, but it needs to get fixed either way.

**Comment by noot** [comment]
okay cool, let me know if I need to make any changes.

**Comment by noot** [comment]
@jacobheun hey, any updates or changes needed?

**Comment by jacobheun** [comment]
@noot not yet, I was away last week, hoping to take a look at it this week.

File: test/pubsub.node.js [hunk 4]
```diff
@@ -52,26 +53,37 @@ function stopTwo (nodes, callback) {
...
+        (cb) => startTwo((err, _nodes) => {
+          nodes = _nodes
+          cb(err)
+        }),
+        // subscribe on the first
+        (cb) => nodes[0].pubsub.subscribe('pubsub', handler, cb),
+        // Wait a moment before publishing
+        (cb) => setTimeout(cb, 500),
+        // publish on the second
+        (cb) => nodes[1].pubsub.publish('pubsub', data, cb),
```
Comment (jacobheun) [inline test/pubsub.node.js]: @noot sorry for the delay, it looks like the tests were just failing because windows needs some time before unsubscribing. This should fix it:
```suggestion
        (cb) => nodes[1].pubsub.publish('pubsub', data, cb),
        // Wait a moment before unsubscribing
        (cb) => setTimeout(cb, 500),
```

Comment (noot) [reply test/pubsub.node.js]: I've made the changes!

**Comment by jacobheun** [comment]
@noot thank you for the PR! I've squashed and merged this to master via 679d446daa0b43c5a1d77c8f300b34c32c678022 to fix the code and commitlint.
